src/ppx/grfx/vk/vk_device.cpp

```diff
@@ -243,6 +261,168 @@ Result Device::ConfigureFeatures(const grfx::DeviceCreateInfo* pCreateInfo, VkPh
+    const grfx::DeviceCreateInfo*                           pCreateInfo,
+    const VkPhysicalDeviceFragmentShadingRateFeaturesKHR&   vrsFeatures,
+    const VkPhysicalDeviceFragmentShadingRatePropertiesKHR& vrsProperties,
+    grfx::FoveationCapabilities*                            pFoveationCapabilities)
+{
+    vk::Gpu* pGpu = ToApi(pCreateInfo->pGpu);
+
+    PFN_vkGetPhysicalDeviceFragmentShadingRatesKHR
+        mFnGetPhysicalDeviceFragmentShadingRatesKHR =
+            (PFN_vkGetPhysicalDeviceFragmentShadingRatesKHR)vkGetInstanceProcAddr(
+                ToApi(GetInstance())->GetVkInstance(), "vkGetPhysicalDeviceFragmentShadingRatesKHR");
+
+    if (mFnGetPhysicalDeviceFragmentShadingRatesKHR == nullptr) {
+        PPX_LOG_INFO("ConfigureVRSShadingRates: Failed to load vkGetPhysicalDeviceFragmentShadingRatesKHR");
+        pFoveationCapabilities->vrs = {};
+        return;
+    }
+
+    uint32_t& supportedRateCount = pFoveationCapabilities->vrs.vk.supportedRateCount;
+    uint8_t*  supportedRates     = pFoveationCapabilities->vrs.vk.supportedRates;
+    uint8_t*  mapRateToSupported = pFoveationCapabilities->vrs.vk.mapRateToSupported;
+
+    VkResult vkres = mFnGetPhysicalDeviceFragmentShadingRatesKHR(pGpu->GetVkGpu(), &supportedRateCount, nullptr);
+
+    std::vector<VkPhysicalDeviceFragmentShadingRateKHR> fragmentShadingRates(
+        supportedRateCount, VkPhysicalDeviceFragmentShadingRateKHR{VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FRAGMENT_SHADING_RATE_KHR});
+    vkres = mFnGetPhysicalDeviceFragmentShadingRatesKHR(pGpu->GetVkGpu(), &supportedRateCount, fragmentShadingRates.data());
+
+    for (uint32_t i = 0; i < 3; ++i) {
+        for (uint32_t j = 0; j < 3; ++j) {
+            uint32_t encoded            = (i << 2) | j;
+            mapRateToSupported[encoded] = UINT8_MAX;
+        }
+    }
+
+    for (uint32_t i = 0; i < supportedRateCount; ++i) {
+        const auto& rate            = fragmentShadingRates[i];
+        uint32_t    w               = rate.fragmentSize.width;
+        uint32_t    h               = rate.fragmentSize.height;
+        uint8_t     encoded         = ((w >> 1) << 2) | (h >> 1);
+        supportedRates[i]           = encoded;
+        mapRateToSupported[encoded] = encoded;
+    }
+
+    for (uint32_t i = 0; i < 3; ++i) {
+        for (uint32_t j = 0; j < 3; ++j) {
+            uint32_t encoded      = (i << 2) | j;
+            uint32_t shrinkWidth  = ((i - 1) << 2) | j;
+            uint32_t shrinkHeight = (i << 2) | (j - 1);
+            if (mapRateToSupported[encoded] == UINT8_MAX) {
+                if (i == 0) {
+                    mapRateToSupported[encoded] = mapRateToSupported[shrinkHeight];
+                }
+                else if (j == 0) {
+                    mapRateToSupported[encoded] = mapRateToSupported[shrinkWidth];
+                }
+                else {
+                    mapRateToSupported[encoded] = std::max<uint8_t>(mapRateToSupported[shrinkWidth], mapRateToSupported[shrinkHeight]);
```

Comment:
nit: The max here is a bit confusing, this line almost always picks shrinkHeight due to encoding.

Reply:
I think the max is necessary. E.g. assume the supported shading rates are 1x1 (encoded=0), 2x2 (encoded=5), 4x4 (encoded=9). Then 4x2 should map to 2x2, but shrinkHeight is 4x1, which maps to 1x1.

Reply:
I'm not claiming it's unnecessary, but it can be confusing on the substitution criteria (therefore it's a nit).
E.g. with available rate 1x1, 2x1, 4x1, 1x2, 1x4, 2x2 the max strategy will pick 4x1 instead rather than 2x2 for 4x4, although I don't think any hardware have such weird set of rates.

Reply:
I added a comment clarifying the behaviour here. That will be in the follow-up PR that handles filling in the shading rate images.